File: src/com/intellij/idea/plugin/hybris/project/configurators/impl/DefaultLibRootsConfigurator.java
/*
 * This file is part of "hybris integration" plugin for Intellij IDEA.
 * Copyright (C) 2014-2016 Alexander Bartash <AlexanderBartash@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.project.configurators.impl;

import com.intellij.idea.plugin.hybris.common.HybrisConstants;
import com.intellij.idea.plugin.hybris.common.LibraryDescriptorType;
import com.intellij.idea.plugin.hybris.project.configurators.LibRootsConfigurator;
import com.intellij.idea.plugin.hybris.project.descriptors.CoreHybrisModuleDescriptor;
import com.intellij.idea.plugin.hybris.project.descriptors.DefaultJavaLibraryDescriptor;
import com.intellij.idea.plugin.hybris.project.descriptors.HybrisModuleDescriptor;
import com.intellij.idea.plugin.hybris.project.descriptors.JavaLibraryDescriptor;
import com.intellij.idea.plugin.hybris.project.descriptors.OotbHybrisModuleDescriptor;
import com.intellij.idea.plugin.hybris.project.descriptors.PlatformHybrisModuleDescriptor;
import com.intellij.idea.plugin.hybris.settings.HybrisApplicationSettings;
import com.intellij.idea.plugin.hybris.settings.HybrisApplicationSettingsComponent;
import com.intellij.openapi.externalSystem.service.project.IdeModifiableModelsProvider;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.roots.DependencyScope;
import com.intellij.openapi.roots.LibraryOrderEntry;
import com.intellij.openapi.roots.ModifiableRootModel;
import com.intellij.openapi.roots.OrderRootType;
import com.intellij.openapi.roots.libraries.Library;
import com.intellij.openapi.roots.libraries.LibraryTable;
import com.intellij.openapi.vfs.JarFileSystem;
import com.intellij.openapi.vfs.VfsUtil;
import com.intellij.openapi.vfs.VirtualFile;
import org.apache.commons.lang3.Validate;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static com.intellij.idea.plugin.hybris.common.HybrisConstants.HYBRIS_PLATFORM_CODE_SERVER_JAR_SUFFIX;

/**
 * Created 11:45 PM 24 June 2015.
 *
 * @author Vlad Bozhenok <VladBozhenok@gmail.com>
 * @author Alexander Bartash <AlexanderBartash@gmail.com>
 */
public class DefaultLibRootsConfigurator implements LibRootsConfigurator {

    @Override
    public void configure(
        @NotNull final ModifiableRootModel modifiableRootModel,
        @NotNull final HybrisModuleDescriptor moduleDescriptor,
        @NotNull IdeModifiableModelsProvider modifiableModelsProvider,
        @NotNull final ProgressIndicator indicator
    ) {
        final VirtualFile sourceCodeRoot = this.getSourceCodeRoot(moduleDescriptor);

        for (JavaLibraryDescriptor javaLibraryDescriptor : moduleDescriptor.getLibraryDescriptors()) {
            if (!javaLibraryDescriptor.isValid() && javaLibraryDescriptor.getScope() == DependencyScope.COMPILE) {
                continue;
            }
            if (javaLibraryDescriptor.isDirectoryWithClasses()) {
                this.addClassesToModuleLibs(
                    modifiableRootModel,
                    modifiableModelsProvider,
                    sourceCodeRoot,
                    javaLibraryDescriptor,
                    moduleDescriptor,
                    indicator
                );
            } else {
                this.addJarFolderToModuleLibs(
                    modifiableRootModel,
                    modifiableModelsProvider,
                    sourceCodeRoot,
                    javaLibraryDescriptor,
                    moduleDescriptor,
                    indicator
                );

            }
        }

        if (moduleDescriptor instanceof PlatformHybrisModuleDescriptor) {
            final PlatformHybrisModuleDescriptor hybrisModuleDescriptor = (PlatformHybrisModuleDescriptor) moduleDescriptor;
            hybrisModuleDescriptor.createBootstrapLib(sourceCodeRoot, modifiableModelsProvider);
        }

        if (moduleDescriptor instanceof CoreHybrisModuleDescriptor) {
            addLibsToModule(
                modifiableRootModel,
                modifiableModelsProvider,
                HybrisConstants.PLATFORM_LIBRARY_GROUP,
                true
            );
        }

        if (moduleDescriptor instanceof OotbHybrisModuleDescriptor) {
            final OotbHybrisModuleDescriptor hybrisModuleDescriptor = (OotbHybrisModuleDescriptor) moduleDescriptor;
            if (hybrisModuleDescriptor.hasBackofficeModule()) {
                final File backofficeJarDirectory = new File(
                    hybrisModuleDescriptor.getRootDirectory(),
                    HybrisConstants.BACKOFFICE_JAR_DIRECTORY
                );
                if (backofficeJarDirectory.exists()) {
                    hybrisModuleDescriptor.createGlobalLibrary(
                        modifiableModelsProvider,
                        backofficeJarDirectory,
                        HybrisConstants.BACKOFFICE_LIBRARY_GROUP
                    );
                }
            }
            if (moduleDescriptor.getName().equals(HybrisConstants.EXTENSION_NAME_BACK_OFFICE)) {
                addLibsToModule(
                    modifiableRootModel,
                    modifiableModelsProvider,
                    HybrisConstants.BACKOFFICE_LIBRARY_GROUP,
                    true
                );
            }
        }
    }

    @Nullable
    private VirtualFile getSourceCodeRoot(final @NotNull HybrisModuleDescriptor moduleDescriptor) {
        final VirtualFile sourceCodeRoot;
        final File sourceCodeFile = moduleDescriptor.getRootProjectDescriptor().getSourceCodeFile();

        if (null != sourceCodeFile) {
            final VirtualFile sourceVFile = VfsUtil.findFileByIoFile(sourceCodeFile, true);
            if (null == sourceVFile) {
                sourceCodeRoot = null;
            } else if (sourceVFile.isDirectory()) {
                sourceCodeRoot = sourceVFile;
            } else {
                sourceCodeRoot = JarFileSystem.getInstance().getJarRootForLocalFile(sourceVFile);
            }
        } else {
            sourceCodeRoot = null;
        }

        return sourceCodeRoot;
    }

    private void addClassesToModuleLibs(
        @NotNull final ModifiableRootModel modifiableRootModel,
        @NotNull final IdeModifiableModelsProvider modifiableModelsProvider,
        @Nullable final VirtualFile sourceCodeRoot,
        @NotNull final JavaLibraryDescriptor javaLibraryDescriptor,
        @NotNull final HybrisModuleDescriptor moduleDescriptor,
        @NotNull final ProgressIndicator progressIndicator
    ) {
        final Library library = modifiableRootModel.getModuleLibraryTable().createLibrary();
        final Library.ModifiableModel libraryModifiableModel = modifiableModelsProvider
            .getModifiableLibraryModel(library);
        libraryModifiableModel.addRoot(
            VfsUtil.getUrlForLibraryRoot(javaLibraryDescriptor.getLibraryFile()), OrderRootType.CLASSES
        );

        boolean sourceDirAttached = false;
        if (null != javaLibraryDescriptor.getSourcesFile()) {
            final VirtualFile srcDirVF = VfsUtil.findFileByIoFile(javaLibraryDescriptor.getSourcesFile(), true);
            if (null != srcDirVF) {
                libraryModifiableModel.addRoot(srcDirVF, OrderRootType.SOURCES);
                sourceDirAttached = true;
            }
        }

        if (sourceCodeRoot != null && !sourceDirAttached && javaLibraryDescriptor.getLibraryFile().getName().endsWith(
            HYBRIS_PLATFORM_CODE_SERVER_JAR_SUFFIX)) {
            libraryModifiableModel.addRoot(sourceCodeRoot, OrderRootType.SOURCES);
        }


        if (javaLibraryDescriptor.isExported()) {
            this.setLibraryEntryExported(modifiableRootModel, library);
        }
        setLibraryEntryScope(modifiableRootModel, library, javaLibraryDescriptor.getScope());
    }

    private void addJarFolderToModuleLibs(
        @NotNull final ModifiableRootModel modifiableRootModel,
        @NotNull final IdeModifiableModelsProvider modifiableModelsProvider,
        @Nullable final VirtualFile sourceCodeRoot,
        @NotNull final JavaLibraryDescriptor javaLibraryDescriptor,
        @NotNull final HybrisModuleDescriptor moduleDescriptor,
        @NotNull final ProgressIndicator progressIndicator
    ) {
        final LibraryTable projectLibraryTable = modifiableRootModel.getModuleLibraryTable();

        final Library library = projectLibraryTable.createLibrary();
        final Library.ModifiableModel libraryModifiableModel = modifiableModelsProvider
            .getModifiableLibraryModel(library);

        libraryModifiableModel.addJarDirectory(
            VfsUtil.getUrlForLibraryRoot(javaLibraryDescriptor.getLibraryFile()), true
        );

        if (null != javaLibraryDescriptor.getSourcesFile()) {
            final VirtualFile srcDirVF = VfsUtil.findFileByIoFile(javaLibraryDescriptor.getSourcesFile(), true);
            if (null != srcDirVF) {
                libraryModifiableModel.addRoot(srcDirVF, OrderRootType.SOURCES);
            }
        }

        if (javaLibraryDescriptor.isExported()) {
            this.setLibraryEntryExported(modifiableRootModel, library);
        }
        setLibraryEntryScope(modifiableRootModel, library, javaLibraryDescriptor.getScope());

        final List<String> mavenSources = resolveMavenSources(
            modifiableRootModel,
            javaLibraryDescriptor,
            moduleDescriptor,
            progressIndicator
        );

        final List<String> standardSources = resolveStandardProvidedSources(
            javaLibraryDescriptor,
            moduleDescriptor
        );

        final List<String> resultLibs = new ArrayList<>(mavenSources);
        resultLibs.addAll(standardSources);

        for (String resultLib : resultLibs) {
            libraryModifiableModel.addRoot("jar://" + resultLib + "!/", OrderRootType.SOURCES);
        }
    }

    public static List<String> resolveMavenSources(
        final @NotNull ModifiableRootModel modifiableRootModel,
        final @NotNull JavaLibraryDescriptor javaLibraryDescriptor,
        final @NotNull HybrisModuleDescriptor moduleDescriptor,
        final @NotNull ProgressIndicator progressIndicator
    ) {
        if (javaLibraryDescriptor instanceof DefaultJavaLibraryDescriptor) {
            final var defaultJavaLibraryDescriptor = (DefaultJavaLibraryDescriptor) javaLibraryDescriptor;
            if (LibraryDescriptorType.LIB == defaultJavaLibraryDescriptor.getDescriptorType()) {
                return MavenUtils.resolveMavenSources(modifiableRootModel, moduleDescriptor, progressIndicator);
            }
        }
        return Collections.emptyList();
    }

    public static List<String> resolveStandardProvidedSources(
        final @NotNull JavaLibraryDescriptor javaLibraryDescriptor,
        final @NotNull HybrisModuleDescriptor moduleDescriptor
    ) {
        final HybrisApplicationSettings appSettings = HybrisApplicationSettingsComponent.getInstance().getState();
        if (!appSettings.isWithStandardProvidedSources()) {
            return Collections.emptyList();
        }

        if (javaLibraryDescriptor instanceof DefaultJavaLibraryDescriptor) {
            final var defaultJavaLibraryDescriptor = (DefaultJavaLibraryDescriptor) javaLibraryDescriptor;
            if (LibraryDescriptorType.WEB_INF_LIB == defaultJavaLibraryDescriptor.getDescriptorType()) {
                var sourcesDirectory = new File(
                    moduleDescriptor.getRootDirectory(),
                    HybrisConstants.DOC_SOURCES_JAR_DIRECTORY
                );

                final String[] filesArray = sourcesDirectory.list((file, name) -> name.endsWith("-sources.jar"));
                if (filesArray == null) {
                    return Collections.emptyList();
                }
                return Arrays.stream(filesArray)
                             .map(fileName -> new File(sourcesDirectory,fileName).getAbsolutePath())
                             .collect(Collectors.toList());
            }
        }
        return Collections.emptyList();
    }

    private void addLibsToModule(
        @NotNull final ModifiableRootModel modifiableRootModel,
        @NotNull IdeModifiableModelsProvider modifiableModelsProvider,
        @NotNull final String libraryName,
        final boolean export
    ) {
        Validate.notNull(modifiableRootModel);

        final LibraryTable.ModifiableModel libraryTableModifiableModel = modifiableModelsProvider
            .getModifiableProjectLibrariesModel();

        Library library = libraryTableModifiableModel.getLibraryByName(libraryName);

        if (null == library) {
            library = libraryTableModifiableModel.createLibrary(libraryName);
        }
        modifiableRootModel.addLibraryEntry(library);

        if (export) {
            setLibraryEntryExported(modifiableRootModel, library);
        }
    }

    private void setLibraryEntryExported(
        @NotNull final ModifiableRootModel modifiableRootModel,
        @NotNull final Library library
    ) {
        final LibraryOrderEntry libraryOrderEntry = findOrderEntryForLibrary(modifiableRootModel, library);

        if (libraryOrderEntry != null) {
            libraryOrderEntry.setExported(true);
        }
    }

    private void setLibraryEntryScope(
        @NotNull final ModifiableRootModel modifiableRootModel,
        @NotNull final Library library,
        @NotNull DependencyScope scope
    ) {
        final LibraryOrderEntry entry = findOrderEntryForLibrary(modifiableRootModel, library);

        if (entry != null) {
            entry.setScope(scope);
        }
    }

    private static LibraryOrderEntry findOrderEntryForLibrary(
        final @NotNull ModifiableRootModel modifiableRootModel,
        final @NotNull Library library
    ) {
        // Workaround of using Library.equals in findLibraryOrderEntry, which doesn't work here, because all empty libs are equal. Use == instead.
        return (LibraryOrderEntry) Arrays
            .stream(modifiableRootModel.getOrderEntries())
            .filter(entry -> entry instanceof LibraryOrderEntry && ((LibraryOrderEntry) entry).getLibrary() == library)
            .findFirst()
            .orElse(modifiableRootModel.findLibraryOrderEntry(library));
    }
}

File: src/com/intellij/idea/plugin/hybris/project/configurators/impl/DefaultLibRootsConfigurator.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for IntelliJ IDEA.
 * Copyright (C) 2019-2024 EPAM Systems <hybrisideaplugin@epam.com> and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.intellij.idea.plugin.hybris.project.configurators.impl

import com.intellij.idea.plugin.hybris.common.HybrisConstants
import com.intellij.idea.plugin.hybris.common.utils.HybrisI18NBundleUtils
import com.intellij.idea.plugin.hybris.project.configurators.LibRootsConfigurator
import com.intellij.idea.plugin.hybris.project.descriptors.JavaLibraryDescriptor
import com.intellij.idea.plugin.hybris.project.descriptors.ModuleDescriptor
import com.intellij.idea.plugin.hybris.project.descriptors.YModuleDescriptor
import com.intellij.idea.plugin.hybris.project.descriptors.YModuleLibDescriptorUtil
import com.intellij.idea.plugin.hybris.project.descriptors.impl.PlatformModuleDescriptor
import com.intellij.idea.plugin.hybris.project.descriptors.impl.YCoreExtModuleDescriptor
import com.intellij.idea.plugin.hybris.project.descriptors.impl.YOotbRegularModuleDescriptor
import com.intellij.idea.plugin.hybris.project.descriptors.impl.YWebSubModuleDescriptor
import com.intellij.openapi.externalSystem.service.project.IdeModifiableModelsProvider
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.roots.DependencyScope
import com.intellij.openapi.roots.LibraryOrderEntry
import com.intellij.openapi.roots.ModifiableRootModel
import com.intellij.openapi.roots.OrderRootType
import com.intellij.openapi.roots.libraries.Library
import com.intellij.openapi.vfs.JarFileSystem
import com.intellij.openapi.vfs.VfsUtil
import com.intellij.openapi.vfs.VirtualFile
import java.io.File

class DefaultLibRootsConfigurator : LibRootsConfigurator {

    override fun configure(
        progressIndicator: ProgressIndicator,
        allYModules: Map<String, YModuleDescriptor>,
        modifiableRootModel: ModifiableRootModel,
        moduleDescriptor: ModuleDescriptor,
        modifiableModelsProvider: IdeModifiableModelsProvider,
        indicator: ProgressIndicator
    ) {
        indicator.text2 = HybrisI18NBundleUtils.message("hybris.project.import.module.libs")

        val sourceCodeRoot = getSourceCodeRoot(moduleDescriptor)
        for (javaLibraryDescriptor in YModuleLibDescriptorUtil.getLibraryDescriptors(moduleDescriptor, allYModules)) {
            if (!javaLibraryDescriptor.libraryFile.exists() && javaLibraryDescriptor.scope == DependencyScope.COMPILE) {
                continue
            }
            if (javaLibraryDescriptor.directoryWithClasses) {
                addClassesToModuleLibs(modifiableRootModel, modifiableModelsProvider, sourceCodeRoot, javaLibraryDescriptor)
            } else {
                addJarFolderToModuleLibs(modifiableRootModel, modifiableModelsProvider, javaLibraryDescriptor)
            }
        }

        when (moduleDescriptor) {
            is PlatformModuleDescriptor -> moduleDescriptor.createBootstrapLib(sourceCodeRoot, modifiableModelsProvider)
            is YCoreExtModuleDescriptor -> addLibsToModule(modifiableRootModel, modifiableModelsProvider, HybrisConstants.PLATFORM_LIBRARY_GROUP, true)
            is YOotbRegularModuleDescriptor -> {
                if (moduleDescriptor.hasBackofficeModule) {
                    val backofficeJarDirectory = File(moduleDescriptor.moduleRootDirectory, HybrisConstants.BACKOFFICE_JAR_PATH)
                    if (backofficeJarDirectory.exists()) {
                        YModuleLibDescriptorUtil.addRootProjectLibrary(modifiableModelsProvider, backofficeJarDirectory, HybrisConstants.BACKOFFICE_LIBRARY_GROUP)
                    }
                }
                if (moduleDescriptor.name == HybrisConstants.EXTENSION_NAME_BACK_OFFICE) {
                    addLibsToModule(modifiableRootModel, modifiableModelsProvider, HybrisConstants.BACKOFFICE_LIBRARY_GROUP, true)
                }
            }
            is YWebSubModuleDescriptor -> {
                if (moduleDescriptor.owner.name == HybrisConstants.EXTENSION_NAME_BACK_OFFICE) {
                    val classes = File(moduleDescriptor.moduleRootDirectory, HybrisConstants.WEBROOT_WEBINF_CLASSES_PATH)
                    val library = File(moduleDescriptor.moduleRootDirectory, HybrisConstants.WEBROOT_WEBINF_LIB_PATH)
                    YModuleLibDescriptorUtil.addRootProjectLibrary(modifiableModelsProvider, classes, HybrisConstants.BACKOFFICE_LIBRARY_GROUP, false)
                    YModuleLibDescriptorUtil.addRootProjectLibrary(modifiableModelsProvider, library, HybrisConstants.BACKOFFICE_LIBRARY_GROUP)
                }
            }
        }
    }

    private fun getSourceCodeRoot(moduleDescriptor: ModuleDescriptor) = moduleDescriptor.rootProjectDescriptor.sourceCodeFile
        ?.let { VfsUtil.findFileByIoFile(it, true) }
        ?.let {
            if (it.isDirectory) {
                it
            } else {
                JarFileSystem.getInstance().getJarRootForLocalFile(it)
            }
        }

    private fun addClassesToModuleLibs(
        modifiableRootModel: ModifiableRootModel,
        modifiableModelsProvider: IdeModifiableModelsProvider,
        sourceCodeRoot: VirtualFile?,
        javaLibraryDescriptor: JavaLibraryDescriptor
    ) {
        val library = javaLibraryDescriptor.name
            ?.let { modifiableRootModel.moduleLibraryTable.createLibrary(it) }
            ?: modifiableRootModel.moduleLibraryTable.createLibrary()
        val libraryModifiableModel = modifiableModelsProvider.getModifiableLibraryModel(library)
        libraryModifiableModel.addRoot(VfsUtil.getUrlForLibraryRoot(javaLibraryDescriptor.libraryFile), OrderRootType.CLASSES)

        val sourceDirAttached = attachSourceFiles(javaLibraryDescriptor, libraryModifiableModel).isNotEmpty()
        attachSourceJarDirectories(javaLibraryDescriptor, libraryModifiableModel)

        if (sourceCodeRoot != null
            && !sourceDirAttached
            && javaLibraryDescriptor.libraryFile.name.endsWith(HybrisConstants.HYBRIS_PLATFORM_CODE_SERVER_JAR_SUFFIX)
        ) {
            libraryModifiableModel.addRoot(sourceCodeRoot, OrderRootType.SOURCES)
        }

        if (javaLibraryDescriptor.exported) {
            setLibraryEntryExported(modifiableRootModel, library)
        }

        setLibraryEntryScope(modifiableRootModel, library, javaLibraryDescriptor.scope)
    }

    private fun addJarFolderToModuleLibs(
        modifiableRootModel: ModifiableRootModel,
        modifiableModelsProvider: IdeModifiableModelsProvider,
        javaLibraryDescriptor: JavaLibraryDescriptor
    ) {
        val projectLibraryTable = modifiableRootModel.moduleLibraryTable
        val library = javaLibraryDescriptor.name
            ?.let { projectLibraryTable.createLibrary(it) }
            ?: projectLibraryTable.createLibrary()

        val libraryModifiableModel = modifiableModelsProvider.getModifiableLibraryModel(library)
        libraryModifiableModel.addJarDirectory(VfsUtil.getUrlForLibraryRoot(javaLibraryDescriptor.libraryFile), true)
        // we have to add each jar file explicitly, otherwise Spring will not recognise `classpath:/META-INF/my.xml` in the jar files
        // Jetbrains IntelliJ IDEA issue - https://youtrack.jetbrains.com/issue/IDEA-257819
        javaLibraryDescriptor.jarFiles.forEach {
            libraryModifiableModel.addRoot(VfsUtil.getUrlForLibraryRoot(it), OrderRootType.CLASSES)
        }

        attachSourceFiles(javaLibraryDescriptor, libraryModifiableModel)
        attachSourceJarDirectories(javaLibraryDescriptor, libraryModifiableModel)

        if (javaLibraryDescriptor.exported) {
            setLibraryEntryExported(modifiableRootModel, library)
        }

        setLibraryEntryScope(modifiableRootModel, library, javaLibraryDescriptor.scope)
    }

    private fun addLibsToModule(
        modifiableRootModel: ModifiableRootModel,
        modifiableModelsProvider: IdeModifiableModelsProvider,
        libraryName: String,
        export: Boolean
    ) {
        val libraryTableModifiableModel = modifiableModelsProvider.modifiableProjectLibrariesModel
        val library = libraryTableModifiableModel.getLibraryByName(libraryName)
            ?: libraryTableModifiableModel.createLibrary(libraryName)
        modifiableRootModel.addLibraryEntry(library)

        if (export) {
            setLibraryEntryExported(modifiableRootModel, library)
        }
    }

    private fun setLibraryEntryExported(modifiableRootModel: ModifiableRootModel, library: Library) {
        findOrderEntryForLibrary(modifiableRootModel, library).isExported = true
    }

    private fun setLibraryEntryScope(modifiableRootModel: ModifiableRootModel, library: Library, scope: DependencyScope) {
        findOrderEntryForLibrary(modifiableRootModel, library).scope = scope
    }

    private fun attachSourceFiles(
        javaLibraryDescriptor: JavaLibraryDescriptor,
        libraryModifiableModel: Library.ModifiableModel
    ) = javaLibraryDescriptor.sourceFiles
        .mapNotNull { VfsUtil.findFileByIoFile(it, true) }
        .onEach { libraryModifiableModel.addRoot(it, OrderRootType.SOURCES) }

    private fun attachSourceJarDirectories(
        javaLibraryDescriptor: JavaLibraryDescriptor,
        libraryModifiableModel: Library.ModifiableModel
    ) {
        javaLibraryDescriptor.sourceJarDirectories
            .mapNotNull { VfsUtil.findFileByIoFile(it, true) }
            .forEach { libraryModifiableModel.addJarDirectory(it, true, OrderRootType.SOURCES) }
    }

    // Workaround of using Library.equals in findLibraryOrderEntry, which doesn't work here, because all empty libs are equal. Use == instead.
    private fun findOrderEntryForLibrary(
        modifiableRootModel: ModifiableRootModel,
        library: Library
    ) = modifiableRootModel.orderEntries
        .mapNotNull { it as? LibraryOrderEntry }
        .find { it.library == library }
        ?: (modifiableRootModel.findLibraryOrderEntry(library) as LibraryOrderEntry)
}
